pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -687,73 +689,93 @@ private void completeConnect(int clientProtoVersion, String clientVersion, boole
-        // credentials.
-        boolean useOriginalAuthState = (originalAuthState != null);
-        AuthenticationState authState =  useOriginalAuthState ? originalAuthState : this.authState;
-        String authRole = useOriginalAuthState ? originalPrincipal : this.authRole;
-        AuthData brokerData = authState.authenticate(clientData);
-
-        if (log.isDebugEnabled()) {
-            log.debug("Authenticate using original auth state : {}, role = {}", useOriginalAuthState, authRole);
-        }
+        // credentials, but we only can new an authentication state, because some authentication data(TLS, SASL)
+        // based on outside service.
+        // If we can get the role from the authentication sate, the global variable need to be updated.
 
-        if (authState.isComplete()) {
-            // Authentication has completed. It was either:
-            // 1. the 1st time the authentication process was done, in which case we'll send
-            //    a `CommandConnected` response
-            // 2. an authentication refresh, in which case we need to refresh authenticationData
-
-            String newAuthRole = authState.getAuthRole();
-
-            // Refresh the auth data.
-            this.authenticationData = authState.getAuthDataSource();
+        boolean useOriginalAuthState = (originalAuthState != null);
+        if (state == State.Connected) {
+            // For auth challenge, the authentication state requires to be updated.
             if (log.isDebugEnabled()) {
-                log.debug("[{}] Auth data refreshed for role={}", remoteAddress, this.authRole);
+                log.debug("Refreshing authenticate state, original auth state: {}, original auth role: {}, "
+                                + "auth role: {}",
+                        useOriginalAuthState, originalPrincipal, authRole);
             }
-
-            if (!useOriginalAuthState) {
-                this.authRole = newAuthRole;
+            try {
+                if (useOriginalAuthState) {
+                    originalAuthState =
+                            originalAuthenticationProvider.newAuthState(clientData, remoteAddress, sslSession);
+                } else {
+                    authState = authenticationProvider.newAuthState(clientData, remoteAddress, sslSession);
+                }
+            } catch (AuthenticationException e) {
+                return CompletableFuture.failedFuture(e);
             }
+        }
 
-            if (log.isDebugEnabled()) {
-                log.debug("[{}] Client successfully authenticated with {} role {} and originalPrincipal {}",
-                        remoteAddress, authMethod, this.authRole, originalPrincipal);
-            }
+        CompletableFuture<AuthData> authFuture = CompletableFuture.completedFuture(null);
+        if (!useOriginalAuthState && !authState.isComplete()) {
+            authFuture = authState.authenticateAsync(clientData);
```

Comment:
I am starting work on finishing PIP 97 today. I think it might make sense to remove this update to use `authenticateAsync` since it is not related to this PR.

Reply:
This PR is to fix the correct persistence authentication state and data.

Do you want to remove which?

Reply:
We must use the `authenticateAsync`, this is used to check the user permission.

Reply:
I am saying we should not make the switch to be asynchronous in this PR. We should make the switch in a separate PR. I started work this week to complete the PIP 97, which introduced the async changes.

Reply:
Done.

Reply:
Ping @michaeljmarshall